File: src/2025/day9.ts

class Coord {
    line: string
    x: number
    y: number

    constructor (line: string) {
        this.line = line;
        [this.x, this.y] = this._parse(line);
    }

    private _parse(line: string): number[] {
        return line.split(",").map(c => {
            return parseFloat(c)
        })
    }

    public get_size(other: Coord): number {
        const dif_x = Math.abs(this.x - other.x) + 1
        const dif_y = Math.abs(this.y - other.y) + 1
        return dif_x * dif_y
    }
}

class Edge {
    x1: number
    x2: number
    y1: number
    y2: number

    constructor (x1: number, y1: number, x2: number, y2: number) {
        this.x1 = x1;
        this.y1 = y1;
        this.x2 = x2;
        this.y2 = y2;
    }
}


const solve_day_9 = (input: string): number => {
    const list_coord: Coord[] = input.split("\n")
    .filter(l => l.trim() != "")
    .map(l => new Coord(l));

    console.log("Red tiles:", list_coord.length);

    const part_1 = solve_part_1(list_coord);
    console.log("Part 1 - Max area:", part_1);
    console.log("Part 1 check:", part_1 == 4759531084);

    const part_2 = solve_part_2(list_coord);
    console.log("Part 2 - Max area (red/green only):", part_2);
    console.log("Part 2 check :", part_2 == 1539238860)

    return part_2;
}

export default solve_day_9;


function solve_part_1(list_coord: Coord[]): number {
    const len_list = list_coord.length;

    let max: number = 0;

    for (let idx_1 = 0; idx_1 < len_list; idx_1++) {
        for (let idx_2 = idx_1+1; idx_2 < len_list; idx_2++){
            const corner_1 = list_coord[idx_1];
            const corner_2 = list_coord[idx_2];
            const size = corner_1.get_size(corner_2);
            if (size > max) max = size;
        }

    }

    return max
}

function solve_part_2(list_coord: Coord[]): number {
    // Construire les arêtes du polygone
    const edges: Edge[] = [];

    for (let i = 0; i < list_coord.length - 1; i++) {
        edges.push(new Edge(
            list_coord[i].x,
            list_coord[i].y,
            list_coord[i + 1].x,
            list_coord[i + 1].y
        ));
    }

    // Fermer le polygone
    const last_idx = list_coord.length - 1;
    edges.push(new Edge(
        list_coord[last_idx].x,
        list_coord[last_idx].y,
        list_coord[0].x,
        list_coord[0].y
    ));

    const len_list = list_coord.length;

    // Générer toutes les paires avec leur aire potentielle
    interface Pair {
        idx_1: number;
        idx_2: number;
        area: number;
    }

    const pairs: Pair[] = [];
    for (let idx_1 = 0; idx_1 < len_list; idx_1++) {
        for (let idx_2 = idx_1 + 1; idx_2 < len_list; idx_2++) {
            const from_tile = list_coord[idx_1];
            const to_tile = list_coord[idx_2];
            const area = from_tile.get_size(to_tile);
            pairs.push({ idx_1, idx_2, area });
        }
    }

    // Trier par aire décroissante
    pairs.sort((a, b) => b.area - a.area);

    console.log(`Vérification des ${pairs.length} paires de rectangles...`);

    let max_found = 0;

    for (const pair of pairs) {
        // Early exit : si l'aire potentielle est <= au max trouvé, ignorer
        if (pair.area <= max_found) {
            break; // Comme les paires sont triées, on peut sortir complètement
        }

        const from_tile = list_coord[pair.idx_1];
        const to_tile = list_coord[pair.idx_2];

        const min_x = Math.min(from_tile.x, to_tile.x);
        const max_x = Math.max(from_tile.x, to_tile.x);
        const min_y = Math.min(from_tile.y, to_tile.y);
        const max_y = Math.max(from_tile.y, to_tile.y);

        // Optimisation : vérifier la distance de Manhattan au carré
        const manhattan = manhattan_distance(from_tile, to_tile);
        if (manhattan * manhattan <= max_found) {
            continue;
        }

        // Vérifier si le rectangle intersecte avec les arêtes
        if (!rectangles_intersects_edges(min_x, min_y, max_x, max_y, edges)) {
            max_found = pair.area;
        }
    }

    return max_found;
}


function rectangles_intersects_edges(min_x: number, min_y: number, max_x: number, max_y: number, edges: Edge[]): boolean {
    for (const edge of edges) {
        const i_min_x = Math.min(edge.x1, edge.x2);
        const i_max_x = Math.max(edge.x1, edge.x2);
        const i_min_y = Math.min(edge.y1, edge.y2);
        const i_max_y = Math.max(edge.y1, edge.y2);

        // Vérifie si les bounding boxes se chevauchent
        if (min_x < i_max_x && max_x > i_min_x && min_y < i_max_y && max_y > i_min_y) {
            return true;
        }
    }
    return false;
}

function manhattan_distance(a: Coord, b: Coord): number {
    return Math.abs(a.x - b.x) + Math.abs(a.y - b.y);
}

File: srcr/src/days_2025/day9.rs
use std::sync::Arc;
use rayon::prelude::*;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
struct Coord {
    x: i64,
    y: i64,
}

impl Coord {
    fn new(input: &str) -> Self {
        let res = input.split(",")
        .map(|n| {
            n.parse::<i64>().unwrap()
        }).collect::<Vec<i64>>();

        if res.len() != 2 {
            println!("{}", input);
            panic!("Invalide Coordonnées !");
        }
        Coord { x: res[0], y: res[1] }
    }

    fn get_size(&self, other: &Coord) -> i64 {
        let dif_x = (self.x - other.x).abs() + 1;
        let dif_y = (self.y - other.y).abs() + 1;
        dif_x * dif_y
    }
}

#[derive(Debug, Clone, Copy)]
struct Edge {
    x1: i64,
    y1: i64,
    x2: i64,
    y2: i64,
}

// Vérifie si un rectangle intersecte avec les arêtes du polygone
fn rectangles_intersects_edges(min_x: i64, min_y: i64, max_x: i64, max_y: i64, edges: &[Edge]) -> bool {
    for edge in edges {
        let i_min_x = edge.x1.min(edge.x2);
        let i_max_x = edge.x1.max(edge.x2);
        let i_min_y = edge.y1.min(edge.y2);
        let i_max_y = edge.y1.max(edge.y2);

        // Vérifie si les bounding boxes se chevauchent
        if min_x < i_max_x && max_x > i_min_x && min_y < i_max_y && max_y > i_min_y {
            return true;
        }
    }
    false
}

// Calcule la distance de Manhattan entre deux points
fn manhattan_distance(a: &Coord, b: &Coord) -> i64 {
    (a.x - b.x).abs() + (a.y - b.y).abs()
}

fn solve_part1(red_tiles: &[Coord]) -> i64 {
    let len_list = red_tiles.len();

    // Paralléliser avec rayon
    (0..len_list)
        .into_par_iter()
        .flat_map(|idx_1| {
            (idx_1 + 1..len_list)
                .into_par_iter()
                .map(move |idx_2| {
                    let corner_1 = &red_tiles[idx_1];
                    let corner_2 = &red_tiles[idx_2];
                    corner_1.get_size(corner_2)
                })
        })
        .max()
        .unwrap_or(0)
}

fn solve_part2(red_tiles: &[Coord]) -> i64 {
    // Construire les arêtes du polygone
    let mut edges = Vec::new();
    for i in 0..red_tiles.len() - 1 {
        edges.push(Edge {
            x1: red_tiles[i].x,
            y1: red_tiles[i].y,
            x2: red_tiles[i + 1].x,
            y2: red_tiles[i + 1].y,
        });
    }
    // Fermer le polygone
    let last_idx = red_tiles.len() - 1;
    edges.push(Edge {
        x1: red_tiles[last_idx].x,
        y1: red_tiles[last_idx].y,
        x2: red_tiles[0].x,
        y2: red_tiles[0].y,
    });

    let edges = Arc::new(edges);
    let len_list = red_tiles.len();

    // Générer toutes les paires avec leur aire potentielle
    let mut pairs: Vec<(usize, usize, i64)> = Vec::new();
    for idx_1 in 0..len_list {
        for idx_2 in idx_1 + 1..len_list {
            let from_tile = &red_tiles[idx_1];
            let to_tile = &red_tiles[idx_2];

            let area = from_tile.get_size(to_tile);
            pairs.push((idx_1, idx_2, area));
        }
    }

    // Trier par aire décroissante
    pairs.sort_unstable_by(|a, b| b.2.cmp(&a.2));

    println!("Vérification des {} paires de rectangles...", pairs.len());

    // Utiliser un AtomicI64 pour suivre le maximum trouvé
    use std::sync::atomic::{AtomicI64, Ordering};
    let max_found = Arc::new(AtomicI64::new(0));

    pairs
        .par_iter()
        .for_each(|(idx_1, idx_2, potential_area)| {
            // Early exit : si l'aire potentielle est <= au max trouvé, ignorer
            let current_max = max_found.load(Ordering::Relaxed);
            if *potential_area <= current_max {
                return;
            }

            let from_tile = &red_tiles[*idx_1];
            let to_tile = &red_tiles[*idx_2];

            let min_x = from_tile.x.min(to_tile.x);
            let max_x = from_tile.x.max(to_tile.x);
            let min_y = from_tile.y.min(to_tile.y);
            let max_y = from_tile.y.max(to_tile.y);

            // Optimisation : vérifier la distance de Manhattan au carré
            let manhattan = manhattan_distance(from_tile, to_tile);
            if manhattan * manhattan <= current_max {
                return;
            }

            // Vérifier si le rectangle intersecte avec les arêtes
            if !rectangles_intersects_edges(min_x, min_y, max_x, max_y, &edges) {
                // Mettre à jour le maximum
                let mut old_max = max_found.load(Ordering::Relaxed);
                while *potential_area > old_max {
                    match max_found.compare_exchange_weak(
                        old_max,
                        *potential_area,
                        Ordering::Relaxed,
                        Ordering::Relaxed,
                    ) {
                        Ok(_) => break,
                        Err(x) => old_max = x,
                    }
                }
            }
        });

    max_found.load(Ordering::Relaxed)
}

pub fn solve_day_9(input: &str) {

    let list_pos = input.lines()
        .filter_map(|l| {
            if !l.is_empty() {
                Some(Coord::new(l))
            } else {
                None
            }
        })
        .collect::<Vec<Coord>>();

    println!("Red tiles: {:?}\n", list_pos);

    let part1 = solve_part1(&list_pos);
    println!("Part 1 - Max area: {}", part1);

    let part2 = solve_part2(&list_pos);
    println!("Part 2 - Max area (red/green only): {}", part2);
}